### DISCUSSION TITLE
Use a smaller Yjs data structure (e.g., Y.XmlFragment) in editor binding

### DISCUSSION THREAD

**Discussion by lihebi** [facebook/lexical#5434] [General]
The current yjs binding of Lexical is quite heavy, requiring a full `Y.Doc` *and* a map of `Y.Doc`s in order to build the binding.

https://github.com/facebook/lexical/blob/ac61359777161743e2c647ddfc42b0df61ff5144/packages/lexical-yjs/src/Bindings.ts#L36-L37

This creates potential problems when integrating with apps. For example, in my app, I have multiple editors along with some other collaborative attributes in each document, and I would like to use one Y.Doc in one websocket connection to sync all of them.

For reference, [y-prosemirror](https://github.com/yjs/y-prosemirror) uses only a `Y.XmlFragment` for the binding. [y-monaco](https://github.com/yjs/y-monaco) uses only a Y.Text for the binding.

What do people think about the idea of extracting a smaller binding API using only `Y.XmlFragment` or `Y.XmlElement`?

---
Reference code:

```ts
// https://github.com/yjs/y-prosemirror
import { ySyncPlugin, yCursorPlugin, yUndoPlugin, undo, redo } from 'y-prosemirror'
const yxml = ydocument.get('prosemirror', Y.XmlFragment)

const prosemirrorView = new EditorView(document.querySelector('#editor'), {
  state: EditorState.create({
    schema,
    plugins: [
        ySyncPlugin(yxml),
        yCursorPlugin(provider.awareness),
        yUndoPlugin(),
      ].concat(exampleSetup({ schema }))
  })
})
```

```ts
// https://github.com/yjs/y-monaco
import { MonacoBinding } from 'y-monaco'

const ydoc = new Y.Doc()
const provider = new WebsocketProvider(`ws://localhost:1234`, 'monaco', ydoc)
const ytext = ydoc.getText('monaco')

const editor = monaco.editor.create(document.getElementById('monaco-editor'), {
  value: '', // MonacoBinding overwrites this value with the content of type
  language: "javascript"
})

// Bind Yjs to the editor model
const monacoBinding = new MonacoBinding(ytext, editor.getModel(), new Set([editor]), provider.awareness)
```

**Comment by Alex-ray** [comment]
Underneath the hood the Lexical Yjs package uses the `YXmlText` shared type as the root binding via `CollabElementNode` ([ref](https://github.com/facebook/lexical/blob/ac61359777161743e2c647ddfc42b0df61ff5144/packages/lexical-yjs/src/Bindings.ts#L58C9-L58C20)). You could reuse that logic and create a version of the CollaborationPlugin that binds that shared type to any doc you want i.e. many lexical roots bound to many shared type within on YJS doc.

I do think there is room for improvement with this package to make it more modular and easier to reuse for more custom use cases that would be a low amount of effort. Even just being able to set a configuration for what Yjs key the editor is bound to would be a low effort high value change (right now it's fixed as `root`)
